enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,204 @@
+- The file is used by ovnkube-node pod to start a goroutine that constantly monitors 
+  the CPUs where ovnkube-node itself runs and where OVS daemon run, aligning the latter if 
+  needed. (See [CPU aligning](#cpu-aligning)). The file's existence is the signal
+  for ovnk to enable the feature. Its contents are not relevant.
+- The user can set `network.operator.openshift.io/cluster.spec.enableOvsCPUPinning = false` to
+  disable the feature cluster-wide.
+
+```mermaid
+sequenceDiagram
+    Note right of NodeTuningOperator: via MachineConfig
+  
+    NodeTuningOperator->>ovs-vswitchd: set_slice
+    NodeTuningOperator->>ovsdb-server: set_slice
+    NodeTuningOperator->>/etc/openvswitch/enable_cpu_pinning: fwrite('1')
+    activate /etc/openvswitch/enable_cpu_pinning
+    ovnkube-node ->> /etc/openvswitch/enable_cpu_pinning: fstat()
+    
+    activate ovnkube-node
+    loop
+    ovnkube-node ->> ovnkube-node: x = sched_getaffinity(x)
+    ovnkube-node ->> ovs-vswitchd: sched_setaffinity(x)
+    ovnkube-node ->> ovsdb-server: sched_setaffinity(x)
+    end
+
+    deactivate ovnkube-node
+    deactivate /etc/openvswitch/enable_cpu_pinning
+```
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+#### CPU aligning
+
+CPU alignment is implemented using system calls SYS_SCHED_SETAFFINITY and SYS_SCHED_GETAFFINITY.
+Pods from `ovnkube-node` daemonset run in the Burstable Quality of Service (QoS) class, hence their CPU 
+affinities are managed by the [kubelet cpu manager](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/).
+
+#### `/etc/openvswitchd/enable_cpu_pinning` file
+
+The file is used to activate the feature on ovnkube-node according to the following logic:
+a. If the file is missing OR the file file is empty, then the feature is turned `OFF`.
+b. If the file exists and its content is not an empty string, then the feature is turned `ON`.
+c. `ovnkube-node` responds to the file modifications (deletion or content update)
+  at runtime, without the need for a restart.
+
+#### Must-Gather enhancement
+
+As the feature is enabled via a fail on the host filesystem, it's useful to
+add `/etc/openvswitchd/enable_cpu_pinning` file to the [gather_network_logs](https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs) script.
+
+### Risks and Mitigations
+
+|Risk|Mitigation|
+|---|---|
+|After NTO enabled the feature, a network performance decrease occurred in some cluster nodes.|The cluster administrator can turn off the feature on specific nodes by connecting through SSH and deleting the activation file. An overriding MachineConfig can be created to achieve the same goal.|
+
+
+
+## Design Details
+
+### Test Plan
+
+Following scenario should be covered by automated tests:
+
+_Given_ an OpenShift cluster with a node N with 10 CPUs (0,1,...,9), configured with 
+- `cpuManagerPolicy = static`
+- `kube-reserved = 0,1,2,3`
```

Comment:
shouldn't this be `system-reserved`? I thought the guaranteed pods would run in a subset of the kube-reserved CPUs
(please correct me if I got this wrong!!)

Reply:
Right, thanks! fixed